dynamiqs/utils/utils.py

```diff
@@ -550,30 +551,76 @@ def todm(x: Tensor) -> Tensor:
-    r"""Returns the overlap (inner product) $\braket{\varphi,\psi}$ between two kets
-    $\ket\varphi$ and $\ket\psi$.
+def braket(x: Tensor, y: Tensor) -> Tensor:
+    r"""Returns the inner product $\braket{\psi|\varphi}$ between two kets.
 
     Args:
-        x _(..., n, 1)_: First ket.
-        y _(..., n, 1)_: Second ket.
+        x _(..., n, 1)_: Left-side ket.
+        y _(..., n, 1)_: Right-side ket.
 
     Returns:
-        _(...)_ Complex-valued overlap.
+        _(...)_ Complex-valued inner product.
 
     Examples:
         >>> fock0 = dq.fock(3, 0)
-        >>> dq.ket_overlap(fock0, fock0)
-        tensor(1.+0.j)
-        >>> fock1 = dq.fock(3, 1)
-        >>> dq.ket_overlap(fock0, fock1)
-        tensor(0.+0.j)
-        >>> alpha0 = dq.coherent(8, 0.0)
-        >>> alpha1 = dq.coherent(8, 1.0)
-        >>> dq.ket_overlap(alpha0, alpha1)
-        tensor(0.607+0.j)
+        >>> fock01 = dq.unit(dq.fock(3, 0) + dq.fock(3, 1))
+        >>> dq.braket(fock0, fock01)
+        tensor(0.707+0.j)
     """
-    return (tobra(x) @ y).squeeze(-1).sum(-1)
+    if not isket(x):
+        raise ValueError(f'Argument `x` must be a ket but has shape {tuple(x.shape)}.')
+    if not isket(y):
+        raise ValueError(f'Argument `y` must be a ket but has shape {tuple(y.shape)}.')
+
+    return (x.mH @ y).squeeze(-2, -1)
+
+
+def overlap(x: Tensor, y: Tensor) -> Tensor:
+    r"""Returns the overlap between two quantum states.
+
+    The overlap is computed
+
+    - as $\lvert\braket{\psi|\varphi}\rvert^2$ if both arguments are kets $\ket\psi$
+      and $\ket\varphi$,
+    - as $\lvert\bra\psi \rho \ket\psi\rvert$ if one argument is a ket $\ket\psi$ and
+      the other is a density matrix $\rho$,
+    - as $\tr{\rho^\dag\sigma}$ if both arguments are density matrices $\rho$ and
+      $\sigma$.
+
+    Args:
+        x _(..., n, 1) or (..., n, n)_: Ket or density matrix.
+        y _(..., n, 1) or (..., n, n)_: Ket or density matrix.
+
+    Returns:
+        _(...)_ Real-valued overlap.
+
+    Examples:
+        >>> fock0 = dq.fock(3, 0)
+        >>> dq.overlap(fock0, fock0)
+        tensor(1.)
+        >>> fock01_dm = 0.5 * (dq.fock_dm(3, 0) + dq.fock_dm(3, 1))
+        >>> dq.overlap(fock0, fock01_dm)
+        tensor(0.500)
+    """
+    if not isket(x) and not isdm(x):
+        raise ValueError(
+            'Argument `x` must be a ket or density matrix, but has shape'
+            f' {tuple(x.shape)}.'
+        )
+    if not isket(y) and not isdm(y):
+        raise ValueError(
+            'Argument `y` must be a ket or density matrix, but has shape'
+            f' {tuple(y.shape)}.'
+        )
+
+    if isket(x) and isket(y):
+        return (x.mH @ y).squeeze(-2, -1).abs().pow(2)
+    elif isket(x) and isdm(y):
+        return (x.mH @ y @ x).squeeze(-2, -1).abs()
+    elif isdm(x) and isket(y):
+        return (y.mH @ x @ y).squeeze(-2, -1).abs()
+    elif isdm(x) and isdm(y):
+        return trace(x.mH @ y).squeeze(-2, -1).real
```

Comment:
I find this more readable but as you pref
```suggestion
    if isket(x) and isket(y):
        return (x.mH @ y).squeeze(-2, -1).abs().pow(2)
    elif isket(x):
        return (x.mH @ y @ x).squeeze(-2, -1).abs()
    elif isket(y):
        return (y.mH @ x @ y).squeeze(-2, -1).abs()
    else:
        return trace(x.mH @ y).squeeze(-2, -1).real
```